### PR TITLE
Support menuconfig and drop duplicated command line options

### PR DESCRIPTION
**Pull request by junjiemao1**
This patch set adds the following features to the make system.

    1. 'make menuconfig' now shows a curses-based TUI for modifying
        configurations.

    2. 'PLATFORM=' or 'RELEASE=' are no longer needed when calling 'make' if a
        configuration exists.

The getting started guide is updated accordingly.

v1 -> v2:

    * Fix 'make oldconfig' overwritting the modified .config with out-dated
      values in config.mk.
    * Rename minimalconfig to savedefconfig.
    * Add link to EPEL in the getting started guide.
    * Split the step using a defconfig from those generating a defconfig in the
      getting started guide.

### CODE REVIEW

**Comment by NanlinXie** [comment]
Can one of the admins verify this patch?

**Comment by junjiemao1** [comment]
@gvancuts I've updated the getting started guide a bit to include the installation of python3 because:
1. menuconfig is python3 only,
2. I do not want users to switch their default python to avoid unintended impacts.

But it may look a bit duplicated since python2 & python3 are both involved here. What do you think of it?

**Comment by gvancuts** [comment]
Hi @junjiemao1 , I agree that with it looks a bit redundant with both python2 and python3. How about we only use python3 (exclusively) and do not care whether the user already has python2. The benefits of this approach in my opinion are:
1. We simplify the instructions and keep it to python3 and pip3 only
2. We always have `menuconfig` available by default

Based on my (limited) testing, the following patch allows us to only rely on python3 and pip3, irrespective of what the default python is or whether python2 is even installed.
```
diff --git a/hypervisor/scripts/kconfig/kconfig.mk b/hypervisor/scripts/kconfig/kconfig.mk
index 69a6196..d6ba1fe 100644
--- a/hypervisor/scripts/kconfig/kconfig.mk
+++ b/hypervisor/scripts/kconfig/kconfig.mk
@@ -5,15 +5,15 @@ HV_CONFIG_MK := include/config.mk
 
 KCONFIG_DIR := $(BASEDIR)/../scripts/kconfig
 
-$(eval $(call check_dep_exec,python))
-$(eval $(call check_dep_exec,pip))
+$(eval $(call check_dep_exec,python3))
+$(eval $(call check_dep_exec,pip3))
 $(eval $(call check_dep_pylib,kconfiglib))
 
 # This target invoke silentoldconfig to generate a .config only if a .config
 # does not exist. Useful as a dependency for source compilation.
 $(HV_OBJDIR)/$(HV_CONFIG):
        @mkdir -p $(HV_OBJDIR)
-       @python $(KCONFIG_DIR)/silentoldconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) PLATFORM_$(shell echo $(PLATFORM) | tr a-z A-Z)=y
+       @python3 $(KCONFIG_DIR)/silentoldconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) PLATFORM_$(shell echo $(PLATFORM) | tr a-z A-Z)=y
 
 $(HV_OBJDIR)/$(HV_CONFIG_MK): $(HV_OBJDIR)/$(HV_CONFIG)
        @mkdir -p $(dir $@)
@@ -21,14 +21,14 @@ $(HV_OBJDIR)/$(HV_CONFIG_MK): $(HV_OBJDIR)/$(HV_CONFIG)
 
 $(HV_OBJDIR)/$(HV_CONFIG_H): $(HV_OBJDIR)/$(HV_CONFIG)
        @mkdir -p $(dir $@)
-       @python $(KCONFIG_DIR)/generate_header.py Kconfig $< $@
+       @python3 $(KCONFIG_DIR)/generate_header.py Kconfig $< $@
 
 # This target forcefully generate a .config based on a given default
 # one. Overwrite the current .config if it exists.
 .PHONY: defconfig
 defconfig:
        @mkdir -p $(HV_OBJDIR)
-       @python $(KCONFIG_DIR)/defconfig.py Kconfig arch/x86/configs/$(PLATFORM).config $(HV_OBJDIR)/$(HV_CONFIG)
+       @python3 $(KCONFIG_DIR)/defconfig.py Kconfig arch/x86/configs/$(PLATFORM).config $(HV_OBJDIR)/$(HV_CONFIG)
 
 # Use silentoldconfig to forcefully update the current .config, or generate a
 # new one if no previous .config exists. This target can be used as a
@@ -37,13 +37,13 @@ defconfig:
 .PHONY: oldconfig
 oldconfig:
        @mkdir -p $(HV_OBJDIR)
-       @python $(KCONFIG_DIR)/silentoldconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) PLATFORM_$(shell echo $(PLATFORM) | tr a-z A-Z)=y
+       @python3 $(KCONFIG_DIR)/silentoldconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) PLATFORM_$(shell echo $(PLATFORM) | tr a-z A-Z)=y
 
 # Minimize the current .config. This target can be used to generate a defconfig
 # for future use.
 .PHONY: minimalconfig
 minimalconfig: $(HV_OBJDIR)/$(HV_CONFIG)
-       @python $(KCONFIG_DIR)/minimalconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) $(HV_OBJDIR)/$(HV_DEFCONFIG)
+       @python3 $(KCONFIG_DIR)/minimalconfig.py Kconfig $(HV_OBJDIR)/$(HV_CONFIG) $(HV_OBJDIR)/$(HV_DEFCONFIG)
 
 -include $(HV_OBJDIR)/$(HV_CONFIG_MK)
 
diff --git a/scripts/deps.mk b/scripts/deps.mk
index a2199f0..1882d27 100644
--- a/scripts/deps.mk
+++ b/scripts/deps.mk
@@ -12,14 +12,14 @@ check_$(1):
         fi
 endef
 
-# check the existence of a specific python library
+# check the existence of a specific python3 library
 # usage: check_dep_pylib <library name>
 define check_dep_pylib =
 BUILD_DEPS += check_$(1)
 check_$(1):
-       @if ! pip list 2>/dev/null | grep $(1) > /dev/null 2>&1; then   \
+       @if ! pip3 list 2>/dev/null | grep $(1) > /dev/null 2>&1; then   \
             echo "******** Missing prerequisite tool ********"; \
-            echo "The python library *$(1)* is not installed"; \
+            echo "The python3 library *$(1)* is not installed"; \
             echo "Please refer to the Getting Started Guide" \
                  "for installation instructions"; \
             exit 1; \
```

What do you think?

**Comment by gvancuts** [comment]
My previous patch does not apply to your tree. Here is the same one (that does apply on top of your working branch (and includes my proposed doc changes):
```
diff --git a/doc/getting_started/index.rst b/doc/getting_started/index.rst
index 348bd8c..925c713 100644
--- a/doc/getting_started/index.rst
+++ b/doc/getting_started/index.rst
@@ -393,9 +393,10 @@ ARCN development is supported on popular Linux distributions,
 each with their own way to install development tools:
 
   .. note::
-     It is recommended to install kconfiglib for both python2 and python3 on
-     distributions using python2 by default because the basic kconfig is
-     supported for both versions but menuconfig is python3-only.
+     We need Python's ``kconfiglib`` to configure the hypervisor. We only
+     support the version for python3 because it is the only one
+     that provides ``menuconfig`` which we use to offer a configuration
+     Text-based User Interface (TUI) to the user.
 
 * On a Clear Linux development system, install the ``os-clr-on-clr`` bundle to get
   the necessary tools:
@@ -419,12 +420,9 @@ each with their own way to install development tools:
           libevent-dev \
           libxml2-dev \
           libusb-1.0-0-dev \
-          python \
-          python-pip \
           python3 \
           python3-pip
-     $ sudo pip install kconfiglib
-     $ sudo pip3 install kconfiglib   # For 'make menuconfig'
+     $ sudo pip3 install kconfiglib
 
   .. note::
      Ubuntu 14.04 requires ``libsystemd-journal-dev`` instead of ``libsystemd-dev``
@@ -443,12 +441,9 @@ each with their own way to install development tools:
           libxml2-devel \
           libevent-devel \
           libusbx-devel \
-          python \
-          python-pip \
           python3 \
           python3-pip
-     $ sudo pip install kconfiglib
-     $ sudo pip3 install kconfiglib   # For 'make menuconfig'
+     $ sudo pip3 install kconfiglib
 
 
 * On a CentOS development system:
@@ -464,12 +459,9 @@ each with their own way to install development tools:
              libxml2-devel \
              libevent-devel \
              libusbx-devel \
-             python \
-             python-pip \
              python34 \
              python34-pip
-     $ sudo pip install kconfiglib
-     $ sudo pip3 install kconfiglib   # For 'make menuconfig'
+     $ sudo pip3 install kconfiglib
 
   .. note::
      You may need to install `EPEL <https://fedoraproject.org/wiki/EPEL>`_ for
@@ -576,8 +568,7 @@ the hypervisor using the updated ``.config``.
       $ make                         # Build the hypervisor with the new .config
 
    .. note::
-      Menuconfig is python3 only. Thus you need to install python3 and
-      kconfiglib for python3.
+      menuconfig is python3 only.
 
 Refer to the help on menuconfig for a detailed guide on the interface.
 
diff --git a/hypervisor/scripts/kconfig/kconfig.mk b/hypervisor/scripts/kconfig/kconfig.mk
index 0b66eed..dc371b5 100644
--- a/hypervisor/scripts/kconfig/kconfig.mk
+++ b/hypervisor/scripts/kconfig/kconfig.mk
@@ -33,9 +33,9 @@ endif
 $(eval $(call override_config,PLATFORM,sbl))
 $(eval $(call override_config,RELEASE,n))
 
-$(eval $(call check_dep_exec,python,BUILD_DEPS))
-$(eval $(call check_dep_exec,pip,BUILD_DEPS))
-$(eval $(call check_dep_pylib,kconfiglib,BUILD_DEPS))
+$(eval $(call check_dep_exec,python3,BUILD_DEPS))
+$(eval $(call check_dep_exec,pip3,BUILD_DEPS))
+$(eval $(call check_dep_py3lib,kconfiglib,BUILD_DEPS))
 
 # This target invoke silentoldconfig to generate or update a .config. Useful as
 # a prerequisite of other targets depending on .config.
@@ -49,14 +49,14 @@ $(HV_OBJDIR)/$(HV_CONFIG_MK): $(HV_OBJDIR)/$(HV_CONFIG)
 
 $(HV_OBJDIR)/$(HV_CONFIG_H): $(HV_OBJDIR)/$(HV_CONFIG)
 	@mkdir -p $(dir $@)
-	@python $(KCONFIG_DIR)/generate_header.py Kconfig $< $@
+	@python3 $(KCONFIG_DIR)/generate_header.py Kconfig $< $@
 
 # This target forcefully generate a .config based on a given default
 # one. Overwrite the current .config if it exists.
 .PHONY: defconfig
 defconfig:
 	@mkdir -p $(HV_OBJDIR)
-	@python $(KCONFIG_DIR)/defconfig.py Kconfig \
+	@python3 $(KCONFIG_DIR)/defconfig.py Kconfig \
 		arch/x86/configs/$(CONFIG_PLATFORM).config \
 		$(HV_OBJDIR)/$(HV_CONFIG)
 
@@ -70,7 +70,7 @@ defconfig:
 .PHONY: oldconfig
 oldconfig:
 	@mkdir -p $(HV_OBJDIR)
-	@python $(KCONFIG_DIR)/silentoldconfig.py Kconfig \
+	@python3 $(KCONFIG_DIR)/silentoldconfig.py Kconfig \
 		$(HV_OBJDIR)/$(HV_CONFIG) \
 		PLATFORM_$(shell echo $(PLATFORM) | tr a-z A-Z)=y \
 		RELEASE=$(RELEASE)
@@ -79,7 +79,7 @@ oldconfig:
 # for future use.
 .PHONY: savedefconfig
 savedefconfig: $(HV_OBJDIR)/$(HV_CONFIG)
-	@python $(KCONFIG_DIR)/savedefconfig.py Kconfig \
+	@python3 $(KCONFIG_DIR)/savedefconfig.py Kconfig \
 		$(HV_OBJDIR)/$(HV_CONFIG) \
 		$(HV_OBJDIR)/$(HV_DEFCONFIG)
```

I'm also thinking that we can most likely completely get rid of the `check_dep_pylib` in `scripts/deps.mk since we only use `check_dep_py3lib` at this point.

**Comment by junjiemao1** [comment]
Thanks for the patch, Geoffroy! I'll combine the changes in this patchset during rebasing and remove the unused check_dep_pylib for now.

File: doc/getting_started/index.rst [hunk 1]
```diff
@@ -392,14 +392,20 @@ Install build tools and dependencies
 ARCN development is supported on popular Linux distributions,
 each with their own way to install development tools:
 
+  .. notes::
```
Comment (dbkinder) [inline doc/getting_started/index.rst]: directive name is misspelled and we can simplify the wording of the note:

    .. note::
       ACRN uses ``menuconfig``, a python3 text-based user interface (TUI) for configuring
       hypervisor options and using python's ``kconfiglib`` library.

File: doc/getting_started/index.rst [hunk 2]
```diff
@@ -513,3 +525,76 @@ and are using it as the current working directory.
 
 Follow the same instructions to boot and test the images you created
 from your build.
+
+Generate the hypervisor configurations
+======================================
+
+The ACRN hypervisor leverages Kconfig to manage configurations, powered by
```
Comment (dbkinder) [inline doc/getting_started/index.rst]: Where do you document the configuration options and choices?

File: doc/getting_started/index.rst (hunk 1 above)
Comment (junjiemao1) [reply doc/getting_started/index.rst]: Will update. Thanks!

File: doc/getting_started/index.rst (hunk 2 above)
Comment (junjiemao1) [reply doc/getting_started/index.rst]: The option documentation should go to the Kconfig files (mainly hypervisor/arch/x86/Kconfig for now). Zephyr has a script that takes a Kconfig and generates a doc in ReST.

File: hypervisor/Makefile [hunk 3]
```diff
@@ -9,23 +9,21 @@ RC_VERSION=5
...
 BASEDIR := $(shell pwd)
-PLATFORM ?= sbl
 HV_OBJDIR ?= $(CURDIR)/build
 HV_FILE := acrn
 
 .PHONY: default
 default: all
 
 include $(BASEDIR)/../scripts/deps.mk
+BUILD_DEPS :=
```
Comment (gvancuts) [inline hypervisor/Makefile]: I'm struggling to follow the entire logic behind this variable but I don't think it's working quite as expected. If `kconfiglib` is missing for example, it still attempts to create a default `.config` using the `oldconfig` target. That fails because `kconfiglib` is not installed. And **then**, it does the check to print out that it's missing. I haven't been able to figure out why it's not running the checks initially but it would be better. Do you have any idea?

File: scripts/deps.mk [hunk 4]
```diff
@@ -12,14 +14,16 @@ check_$(1):
 	 fi
 endef
 
-# check the existence of a specific python library
-# usage: check_dep_pylib <library name>
+# usage: check_dep_pylib <library name> <variable>
+#
+# Create a target that checks the existence of the specified python 3 library, and
+# append that target to the given variable.
 define check_dep_pylib =
```
Comment (gvancuts) [inline scripts/deps.mk]: This function is never called because you use `check_dep_py3lib` above. You should correct the name of the function (or fix the invocation in `scripts/kconfig/kconfig.mk`

File: hypervisor/Makefile (hunk 3 above)
Comment (junjiemao1) [reply hypervisor/Makefile]: ''make'' tries updating all included makefiles (with failures ignored) and re-execute itself if any change happens (with ''make -d'' the logs show this process). Thus what happens here is that:

1. ''make'' first check whether config.mk needs updating. Since config.mk depends on .config and .config on oldconfig, oldconfig in invoked and the absence of kconfiglib is reported by python. This error is ignored by ''make''.
2. Since no included makefiles are modified, ''make'' continues to execute the building target, which depends on the prerequisite checking.

Making the prerequisite checks as dependencies to oldconfig fixes that. I'll update this PR and please kindly verify it when you have time. Thanks!

File: scripts/deps.mk (hunk 4 above)
Comment (junjiemao1) [reply scripts/deps.mk]: Thanks for pointing it out. I'll fix the definition.

**Comment by TianhuaSun** [comment]
Verified+1
